homeassistant/components/vicare/fan.py

```diff
@@ -0,0 +1,192 @@
+]
+
+PRESET_MODES = [
+    VentilationMode.PERMANENT,
+    VentilationMode.VENTILATION,
+    VentilationMode.SENSOR_DRIVEN,
+    VentilationMode.SENSOR_OVERRIDE,
+]
+
+
+def _build_entities(
+    device_list: list[ViCareDevice],
+) -> list[ViCareFan]:
+    """Create ViCare fan entities for a device."""
+
+    return [
+        ViCareFan(device.config, device.api, "ventilation")
+        for device in device_list
+        if isinstance(device.api, PyViCareVentilationDevice)
+    ]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the ViCare fan platform."""
+
+    device_list = hass.data[DOMAIN][config_entry.entry_id][DEVICE_LIST]
+
+    async_add_entities(
+        await hass.async_add_executor_job(
+            _build_entities,
+            device_list,
+        )
+    )
+
+
+class ViCareFan(ViCareEntity, FanEntity):
+    """Representation of the ViCare ventilation device."""
+
+    _attr_speed_count = len(ORDERED_NAMED_FAN_SPEEDS)
+    _attr_supported_features = FanEntityFeature.SET_SPEED | FanEntityFeature.PRESET_MODE
+    _attributes: dict[str, Any] = {
+        "active_vicare_mode": None,
+        "active_vicare_program": None,
+        "vicare_modes": None,
+        "vicare_programs": None,
+    }
+
+    def __init__(
+        self,
+        device_config: PyViCareDeviceConfig,
+        device: PyViCareDevice,
+        translation_key: str,
+    ) -> None:
+        """Initialize the fan entity."""
+        super().__init__(device_config, device, translation_key)
+        self._attr_translation_key = translation_key
+
+        with suppress(PyViCareNotSupportedFeatureError):
+            self._attributes["vicare_modes"] = self._api.getAvailableModes()
+        self._attributes["vicare_programs"] = self._api.getAvailablePrograms()
+
+    def update(self) -> None:
+        """Update state of fan."""
+        try:
+            with suppress(PyViCareNotSupportedFeatureError):
+                self._attributes["active_vicare_mode"] = self._api.getActiveMode()
+            with suppress(PyViCareNotSupportedFeatureError):
+                self._attributes["active_vicare_program"] = self._api.getActiveProgram()
+        except RequestConnectionError:
+            _LOGGER.error("Unable to retrieve data from ViCare server")
+        except ValueError:
+            _LOGGER.error("Unable to decode data from ViCare server")
+        except PyViCareRateLimitError as limit_exception:
+            _LOGGER.error("Vicare API rate limit exceeded: %s", limit_exception)
+        except PyViCareInvalidDataError as invalid_data_exception:
+            _LOGGER.error("Invalid data from Vicare server: %s", invalid_data_exception)
+
+    def turn_off(self, **kwargs: Any) -> None:
+        """Turn the entity off."""
+        raise ServiceValidationError(
+            translation_domain=DOMAIN,
+            translation_key="fan_must_be_on",
+        )
```

Comment:
This is not great, I'd suggest to open an architecture discussion suggesting to add FanEntityFeature.TURN_OFF + FanEntityFeatur.TURN_ON

Reply:
https://github.com/home-assistant/architecture/discussions/1097